modules/stochastic_tools/include/utils/GaussianProcess.h

```diff
@@ -137,6 +137,129 @@ class GaussianProcess
    */
   void standardizeData(RealEigenMatrix & data, bool keep_moments = false);
 
+  // void sq_dist(const RealEigenMatrix &X1_in, RealEigenMatrix &D_out, const RealEigenMatrix &X2_in = RealEigenMatrix(0,0));
+
+  /**
+   * Parameter setting for MCMC sampling.
+   * @param l Constant parameter from the paper.
+   * @param u Constant parameter from the paper.
+   * @param g Noise level.
+   * @param theta Lengthscale.
+   * @param alpha Parameter for gamma distribution.
+   * @param beta Parameter for gamma distribution.
+   */
+  struct Settings {
+    Real l;
+    Real u;
+    struct {
+      Real g;
+      Real theta;
+    } alpha, beta;
+  };
+
+  /**
+   * Initialzed value for GP hyperparameters.
+   * @param theta Lengthscale.
+   * @param g Noise level.
+   * @param tau2 Scale.
+   */
+  struct Initial {
+    RealEigenMatrix theta;
+    Real g;
+    Real tau2;
+  };
+
+  /**
+   * Return value computed in the sample_g function.
+   * @param g Sampled noise level.
+   * @param ll Log likelihood.
+   */
+  struct SampleGResult {
+    Real g;
+    Real ll;
+  };
+
+  /**
+   * Return value computed in the sample_theta function.
+   * @param g Sampled noise level.
+   * @param ll Log likelihood.
+   * @param tau2 Sampled scale.
+   */
+  struct SampleThetaResult {
+    Real theta;
+    Real ll;
+    Real tau2;
+  };
+
+  /**
+   * Return value computed in the logl function.
+   * @param logl Marginal log likelihood.
+   * @param tau2 Scale.
+   */
+  struct LogLResult {
+    Real logl;
+    Real tau2;
+  };
+
+  /**
+   * Computes multivariate normal marginal log likelihood and scale.
+   * @param outvec Observed outputs (response values).
+   * @param x1 Input data 1.
+   * @param x2 Input data 2.
+   * @param g Noise level.
+   * @param theta Lengthscale.
+   * @param result Return value.
+   * @param outer If the function is called for computing outermost layer.
+   * @param tau2 If tau2 is needed to be computed.
+   */
+  void logl(const RealEigenMatrix & out_vec, const RealEigenMatrix & x1, const RealEigenMatrix & x2, Real g, const RealEigenMatrix & theta, 
+          LogLResult & result, bool outer=true, bool tau2=false);
+
+  /**
+   * Samples noise level g using MH algorithm.
+   * @param outvec Observed outputs (response values).
+   * @param x1 Input data 1.
+   * @param x2 Input data 2.
+   * @param g_t Noise level.
+   * @param theta Lengthscale.
+   * @param alpha Parameter for gamma distribution.
+   * @param beta Parameter for gamma distribution.
+   * @param l Constant parameter from the paper.
+   * @param u Constant parameter from the paper.
+   * @param ll_prev Log likelihood from the previous MCMC round.
+   * @param result Return value.
+   */
+  void sample_g(const RealEigenMatrix & out_vec, const RealEigenMatrix & x1, const RealEigenMatrix & x2, Real g_t, const RealEigenMatrix theta, 
```

Comment:
To me it feels like there is no difference in tuning the noise and the scale parameters besides computing the scale parameter. I think we could create one function that can do both of these.